### ISSUE TITLE
Battery icon not keeping up/changing with battery drain

### ISSUE THREAD

**Issue by isantop** [pop-os/pop#54]
_From @kylecorry31 on July 7, 2017 22:25_

**Reporter info**

```
 * Distribution - Pop!_OS
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 17.04
Release:	17.04
Codename:	zesty
 * Related Application and/or Package Version - GNOME Shell, pop icon theme 1.1.10 (not completely sure which application is the related one)
```

**Actual issue**
![image](https://user-images.githubusercontent.com/5938560/27979032-12587e70-6341-11e7-9d95-2fd9474b2575.png)

The battery icon takes a very long time to change its look to indicate how much battery is left. It took between 30 minutes and 45 minutes before the battery icon indicated it was down to around halfway. 

**Steps to reproduce (if you knew)**
Unplug laptop (System76 Gazp9) and allow battery to drain. The icon does not seem to keep up with the draining (reproduced multiple times, but not consistent) 

**Expected behaviour**
This icon should "drain" with the battery percentage to indicate to the user how much is left at a quick glance.

**Other Note (feature-request, question, etc...)**

_Copied from original issue: pop-os/iso#44_

**Comment by isantop** [comment]
_From @kylerchin on July 10, 2017 4:7_

I don't think it is a battery issue. I suggest a bug fix. I have the same problem. I believe the problem may have originated from the gnome shell or Ubuntu's software. To make sure, please post the output of ```upower -d```
Thanks!

**Comment by isantop** [comment]
_From @ryanleesipes on July 10, 2017 21:26_

This is a bug, I also suffer from this. The battery icon does not properly reflect how much life you have left. When it falls below 50% it still looks like > 75%. (Don't have a screenshot yet, have to run my battery down).

**Comment by isantop** [comment]
_From @JosephMcc on July 10, 2017 21:51_

Keep in mind that there are a very limited number of icons to choose from to represent the status. Look in /usr/share/icons/Adwaita/scalable/status. The percentage is much better reflection then the icon representation.

**Comment by isantop** [comment]
_From @kylerchin on July 10, 2017 22:56_

Yeah, but it's still freaking me out. ****shiver****

**Comment by isantop** [comment]
_From @kylecorry31 on July 10, 2017 23:3_

Here is the upower -d output:
[power.txt](https://github.com/system76/pop-distro/files/1137290/power.txt)

UPDATE: here is the upower -d output at 35%, still "battery-good-symbolic":
[power2.txt](https://github.com/system76/pop-distro/files/1137391/power2.txt)


This is what the UI looks like at this point:
![screenshot from 2017-07-10 18-58-55](https://user-images.githubusercontent.com/5938560/28043900-322a29e8-65a2-11e7-9676-244bd5bc1c5c.png)

It looks like the icon being displayed is "battery-good-symbolic", which doesn't seem right because I had around 58% battery. It should probably display the "battery-medium-symbolic" icon at this point. 

I also noticed that the pop-icon-theme has the same icon for both the medium and the low icons, which is a bug in that theme:
![screenshot from 2017-07-10 18-16-48](https://user-images.githubusercontent.com/5938560/28043957-7df26340-65a2-11e7-9488-9fdf860f8b73.png)

I don't know how feasible it would be, but it may be worthwhile to add icons (and support in the system for switching to them of course) for different increments than just full, good, medium, and low. Maybe every 10%?

**Comment by isantop** [comment]
_From @kylerchin on July 11, 2017 1:39_

Let's modify our icons!
![image](https://user-images.githubusercontent.com/7539174/28047472-81446908-65a7-11e7-9100-a5551803d547.png)

**Comment by isantop** [comment]
_From @ryanleesipes on July 28, 2017 19:22_

Marking this as BSB as the icons exist but someone needs to go through and change when they are used.

**Comment by isantop** [comment]
_From @kylecorry31 on July 28, 2017 20:52_

I might take a look at this (I won't have computer access until Tuesday, so if anyone wants to do it before then, feel free to)

**Comment by isantop** [comment]
_From @cduston on July 29, 2017 19:2_

I think this is expected behavior from upower. Taking a look at the source code (https://cgit.freedesktop.org/upower/tree/src/up-device.c#n99) we have these lines for the case of "charging":

get_device_charge_icon (gdouble  percentage,
			gboolean charging)
{
	if (percentage < 10)
		return charging ? "battery-caution-charging-symbolic" : "battery-caution-symbolic";
	else if (percentage < 30)
		return charging ? "battery-low-charging-symbolic" : "battery-low-symbolic";
	else if (percentage < 60)
		return charging ? "battery-good-charging-symbolic" : "battery-good-symbolic";
	return charging ? "battery-full-charging-symbolic" : "battery-full-symbolic";
}

So between 30-60 is "good", according to UPower. I assume the case is similar for "discharging", but I can't quite understand the code in that section. In addition, "medium" is never used (maybe 30-60 should be "medium" and 60+ should be "good"?) In any case, that part is a problem with UPower.

I agree with the modification to the icons though. If I have time I'll just go and erase a little bit of it I suppose...

**Comment by isantop** [comment]
The icons are improved in 287623e0286fc993cdbabb6bd760df8276981378 and 22d7319f24224ee7d7edbfada1916d14cb030ac1

**Comment by gregoryopera** [comment]
Yep, I have this too... Since I installed Pop!_OS a few weeks back, I've frequently found myself doing the mad dash for the charger because I'm down to 30% battery or less and the icon has only just changed to the "50% look".

It really is poorly designed, in its current form.

When my battery is at 50%, the icon should clearly indicate this... Right now, it implies that I still have 75% or more left, which is misleading and frustrating; this is made worse by the fact that I cannot even work around this and display the percentage - I need to actually click the battery to see the percentage.

**Comment by kylecorry31** [comment]
Until this issue is fixed, you could try this: http://support.system76.com/articles/show-battery-percentage/

**Comment by gregoryopera** [comment]
Thanks Kyle, I'll check it out!

**Comment by arjunkrishnababu96** [comment]
I too have this issue (and I just switched from a rock-solid Debian; lol). 

Attached below are screenshots for 31% and 29%.

![bat31](https://user-images.githubusercontent.com/6965563/37133077-54206896-224e-11e8-9acd-a78357ef624d.png)
The picture above looks more like 80%

![bat29](https://user-images.githubusercontent.com/6965563/37133074-52368ee8-224e-11e8-9ba6-affbd14168db.png)


The indicator for 31% and 30% are the same. The switch happens at 29%.

Is anybody working on fixing this?

**Comment by kylecorry31** [comment]
I believe this can be fixed in the gnome shell rather than waiting for UPower to be updated, the following file needs to be modified: https://github.com/GNOME/gnome-shell/blob/master/js/ui/status/power.js, specifically lines 124 - 126

**Comment by kylecorry31** [comment]
I think this fixes it for me: https://github.com/kylecorry31/gnome-shell-extension-pop-battery-icon-fix 

It is a simple, hacked together extension which overrides the _sync method of the power icon to display the icon which was mentioned earlier.

**Comment by isantop** [comment]
Thanks @kylecorry31 ! I'll test it out and see how well it works.

**Comment by sramkrishna** [comment]
It seems we have one more out in the wild too.  It seems this is known in upstream.  https://github.com/Exalm/gnome-shell-extension-dynamic-battery

The problem with this extension is that it breaks themes as the icon is dynamic.

**Comment by isantop** [comment]
@sramkrishna I think we're going to go with @kylecorry31 's. It's pretty much perfect and doesn't break any themes.

**Comment by kylecorry31** [comment]
Yeah, I basically rewrote the icon selection code from UPower in the extension and adjusted the level thresholds to make it match better.

**Comment by sramkrishna** [comment]
That's fine, I'm just iterating on what it is out there.  I think upstream would like something more dynamic, but nothing has materialized in this regard.

**Comment by isantop** [comment]
I would think if upstream wanted something dynamic, they would have implemented it in GNOME Shell, not waited for an extension to do it externally.

**Comment by brs17** [comment]
Closing this issue since https://github.com/pop-os/gnome-shell-extension-pop-battery-icon-fix is now in Pop!_OS 18.04
